### PR TITLE
Refactor Code into Methods, More Versatile Handling of Multiple Devices, Linux Support Improvements

### PR DESCRIPTION
**Pull request by obelisk**
This PR attempts to solve the problem of the library only expecting there to ever be one key connected. It tries to solve this by moving many of the global functions which act upon a device (using the get_device) function into member function of a device.

So far I've moved the the get_info and pin functions into the fidokey module. You can see in the updated get-info example how this is used, and can now fetch the info of all connected keys instead of just the first one the library happens to see.

This also adds build tests for macOS and Ubuntu to ensure that any changes I make at least don't break the build across multiple platforms.

Finally, I've removed the linux specific code and am testing it more extensively on some linux machines that I have to determine exactly what from them is still needed :)

### CODE REVIEW

**Comment by obelisk** [comment]
@gebogebogebo, drop me a line on Keybase @obelisk and we can discuss this PR and future work there more easily if needed!

**Comment by obelisk** [comment]
Confirmed this breaks Linux 😓 

Working on fixing that!

**Comment by obelisk** [comment]
Hello, I'm Linux!
----- get-info start : key_auto = true -----
get_hid_devices()
- vid=0x046d , pid=0x4082 , info="product=USB Receiver usage_page=1 usage=6 serial_number=4082-59-8f-c2-d0 path=\"/dev/hidraw1\""
- vid=0x046d , pid=0x4082 , info="product=USB Receiver usage_page=1 usage=2 serial_number=4082-59-8f-c2-d0 path=\"/dev/hidraw1\""
- vid=0x046d , pid=0x4082 , info="product=USB Receiver usage_page=1 usage=1 serial_number=4082-59-8f-c2-d0 path=\"/dev/hidraw1\""
- vid=0x046d , pid=0x4082 , info="product=USB Receiver usage_page=65280 usage=1 serial_number=4082-59-8f-c2-d0 path=\"/dev/hidraw1\""
- vid=0x046d , pid=0x4082 , info="product=USB Receiver usage_page=65280 usage=2 serial_number=4082-59-8f-c2-d0 path=\"/dev/hidraw1\""
- vid=0x046d , pid=0x4082 , info="product=USB Receiver usage_page=65280 usage=4 serial_number=4082-59-8f-c2-d0 path=\"/dev/hidraw1\""
- vid=0x046d , pid=0xc52b , info="product=USB Receiver usage_page=65280 usage=1 serial_number= path=\"/dev/hidraw0\""
- vid=0x046d , pid=0xc52b , info="product=USB Receiver usage_page=65280 usage=2 serial_number= path=\"/dev/hidraw0\""
- vid=0x046d , pid=0xc52b , info="product=USB Receiver usage_page=65280 usage=4 serial_number= path=\"/dev/hidraw0\""
- vid=0x1b1c , pid=0x0c04 , info="product=Integrated USB Bridge usage_page=65280 usage=1 serial_number= path=\"/dev/hidraw4\""
- vid=0x1050 , pid=0x0407 , info="product=YubiKey OTP+FIDO+CCID usage_page=1 usage=6 serial_number= path=\"/dev/hidraw7\""
- vid=0x1050 , pid=0x0407 , info="product=YubiKey OTP+FIDO+CCID usage_page=61904 usage=1 serial_number= path=\"/dev/hidraw8\""
- vid=0xfeed , pid=0x1307 , info="product=ErgoDox EZ usage_page=1 usage=6 serial_number=0 path=\"/dev/hidraw3\""
- vid=0xfeed , pid=0x1307 , info="product=ErgoDox EZ usage_page=1 usage=128 serial_number=0 path=\"/dev/hidraw5\""
- vid=0xfeed , pid=0x1307 , info="product=ErgoDox EZ usage_page=12 usage=1 serial_number=0 path=\"/dev/hidraw5\""
- vid=0xfeed , pid=0x1307 , info="product=ErgoDox EZ usage_page=1 usage=6 serial_number=0 path=\"/dev/hidraw5\""
get_fidokey_devices()


---------------------------------------------
- vid=0x1050 , pid=0x0407 , info="product=YubiKey OTP+FIDO+CCID usage_page=61904 usage=1 serial_number= path=\"/dev/hidraw8\""
get_info()
- versions                           = ["U2F_V2", "FIDO_2_0", "FIDO_2_1_PRE"]
- extensions                         = ["credProtect", "hmac-secret"]
- aaguid(16)                         = C5EF55FFAD9A4B9FB580ADEBAFE026D0
- options                            = [("rk", true), ("up", true), ("plat", false), ("clientPin", false), ("credentialMgmtPreview", true)]
- max_msg_size                       = 1200
- pin_uv_auth_protocols              = [1]
- max_credential_count_in_list       = 8
- max_credential_id_length           = 128
- transports                         = ["usb", "lightning"]
- algorithms                         = [("alg", "-7"), ("type", "public-key"), ("alg", "-8"), ("type", "public-key")]
- max_serialized_large_blob_array    = 0
- force_pin_change                   = false
- min_pin_length                     = 0
- firmware_version                   = 0
- max_cred_blob_length               = 0
- max_rpids_for_set_min_pin_length   = 0
- preferred_platform_uv_attempts     = 0
- uv_modality                        = 0
- remaining_discoverable_credentials = 0

get_pin_retries()
error: response_status err = 0x35 CTAP2_ERR_PIN_NOT_SET No PIN has been set.
get_info_u2f()
U2F_V2
----- get-info end -----

**Comment by obelisk** [comment]
Ok I think this works now, I changed the backend out for `linux-static-hidraw`. This should give us better parity with macOS, and Windows as well as being able to remove all the Linux specific code!

Here is some output from the new `get-info` build (note that VID and PID are now present):
```
Hello, I'm Linux!
----- get-info start : key_auto = true -----
get_hid_devices()
- vid=0x046d , pid=0x4082 , info="product=USB Receiver usage_page=1 usage=6 serial_number=4082-59-8f-c2-d0 path=\"/dev/hidraw1\""
- vid=0x046d , pid=0x4082 , info="product=USB Receiver usage_page=1 usage=2 serial_number=4082-59-8f-c2-d0 path=\"/dev/hidraw1\""
- vid=0x046d , pid=0x4082 , info="product=USB Receiver usage_page=1 usage=1 serial_number=4082-59-8f-c2-d0 path=\"/dev/hidraw1\""
- vid=0x046d , pid=0x4082 , info="product=USB Receiver usage_page=65280 usage=1 serial_number=4082-59-8f-c2-d0 path=\"/dev/hidraw1\""
- vid=0x046d , pid=0x4082 , info="product=USB Receiver usage_page=65280 usage=2 serial_number=4082-59-8f-c2-d0 path=\"/dev/hidraw1\""
- vid=0x046d , pid=0x4082 , info="product=USB Receiver usage_page=65280 usage=4 serial_number=4082-59-8f-c2-d0 path=\"/dev/hidraw1\""
- vid=0x046d , pid=0xc52b , info="product=USB Receiver usage_page=65280 usage=1 serial_number= path=\"/dev/hidraw0\""
- vid=0x046d , pid=0xc52b , info="product=USB Receiver usage_page=65280 usage=2 serial_number= path=\"/dev/hidraw0\""
- vid=0x046d , pid=0xc52b , info="product=USB Receiver usage_page=65280 usage=4 serial_number= path=\"/dev/hidraw0\""
- vid=0x1b1c , pid=0x0c04 , info="product=Integrated USB Bridge usage_page=65280 usage=1 serial_number= path=\"/dev/hidraw4\""
- vid=0x1050 , pid=0x0407 , info="product=YubiKey OTP+FIDO+CCID usage_page=1 usage=6 serial_number= path=\"/dev/hidraw7\""
- vid=0x1050 , pid=0x0407 , info="product=YubiKey OTP+FIDO+CCID usage_page=61904 usage=1 serial_number= path=\"/dev/hidraw8\""
- vid=0xfeed , pid=0x1307 , info="product=ErgoDox EZ usage_page=1 usage=6 serial_number=0 path=\"/dev/hidraw3\""
- vid=0xfeed , pid=0x1307 , info="product=ErgoDox EZ usage_page=1 usage=128 serial_number=0 path=\"/dev/hidraw5\""
- vid=0xfeed , pid=0x1307 , info="product=ErgoDox EZ usage_page=12 usage=1 serial_number=0 path=\"/dev/hidraw5\""
- vid=0xfeed , pid=0x1307 , info="product=ErgoDox EZ usage_page=1 usage=6 serial_number=0 path=\"/dev/hidraw5\""
get_fidokey_devices()


---------------------------------------------
- vid=0x1050 , pid=0x0407 , info="product=YubiKey OTP+FIDO+CCID usage_page=61904 usage=1 serial_number= path=\"/dev/hidraw8\""
get_info()
- versions                           = ["U2F_V2", "FIDO_2_0", "FIDO_2_1_PRE"]
- extensions                         = ["credProtect", "hmac-secret"]
- aaguid(16)                         = C5EF55FFAD9A4B9FB580ADEBAFE026D0
- options                            = [("rk", true), ("up", true), ("plat", false), ("clientPin", false), ("credentialMgmtPreview", true)]
- max_msg_size                       = 1200
- pin_uv_auth_protocols              = [1]
- max_credential_count_in_list       = 8
- max_credential_id_length           = 128
- transports                         = ["usb", "lightning"]
- algorithms                         = [("alg", "-7"), ("type", "public-key"), ("alg", "-8"), ("type", "public-key")]
- max_serialized_large_blob_array    = 0
- force_pin_change                   = false
- min_pin_length                     = 0
- firmware_version                   = 0
- max_cred_blob_length               = 0
- max_rpids_for_set_min_pin_length   = 0
- preferred_platform_uv_attempts     = 0
- uv_modality                        = 0
- remaining_discoverable_credentials = 0

get_pin_retries()
error: response_status err = 0x35 CTAP2_ERR_PIN_NOT_SET No PIN has been set.
get_info_u2f()
U2F_V2
----- get-info end -----
```

**Comment by gebogebogebo** [comment]
Thank you !
Please give me a moment to take a closer look at the code.

File: .github/workflows/rust.yml [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+    - uses: actions/checkout@v2
+    - name: Build
+      run: cargo build --examples --verbose
+  
+  ubuntu-build:
+    runs-on: ubuntu-latest
+
+    steps:
+    - uses: actions/checkout@v2
+    - name: Build
```
Comment (btielen) [inline .github/workflows/rust.yml:26]: Pretty sure this step will fail because of the issue mentioned in #40

Comment (obelisk) [reply .github/workflows/rust.yml:26]: It passes here at least: https://github.com/obelisk/ctap-hid-fido2/pull/1

I thought I might have to manually install libusb (and this PR also require libudev) but I think they are installed by default in the GitHub images.

**Comment by obelisk** [comment]
Any thoughts on this? Would like some feedback before I jump in and try to refactor more functions.

**Comment by gebogebogebo** [comment]
One PR contains many issues.
Is it possible to divide the discussion into multiple PRs of smaller units?
I find it difficult to make a decision because of the volume of revisions I have to make.

I would like to at least divide it into the following issues
- Handling of Multiple Devices.
- Per-command packaging (refactoring).
- Linux Support Improvements(I do not declare Linux support, I support Raspberry Pi OS)
- adds build tests

How about creating one feature branch and then creating multiple small issues to merge into that branch?

Once you have received consent, create a FEATURE branch.

**Comment by obelisk** [comment]
I can split the PR into a few. Definitely the build tests are easy to split out. Linux support is also easy. The handling of multiple devices though and the per-command packaging are intertwined so they cannot be easily separated.

In addition, the per-command packaging is dependent on the Linux PR so while I can split it, we'd need to merge the linux support first.

As for Raspberry Pi OS support, that is Linux. Raspberry Pi OS is based on Debian Linux, so this will support that, plus others with less code here if we move to the new backend for hidapi on Linux :)

**Comment by gebogebogebo** [comment]
Thank you.
I don't speak English very well.

> How about creating one feature branch and then creating multiple small issues to merge into that branch?
> Once you have received consent, create a FEATURE branch.

What I wanted to say here is that I want you to create a PR that merges into the development branch, not a PR to master.
I just published a [develop] branch.
https://github.com/gebogebogebo/ctap-hid-fido2/tree/develop
Please give PR to this branch. :)

**Comment by obelisk** [comment]
@gebogebogebo Your English is better than my Japanese! You can always message me on Keybase if you need.

I've split the PRs up anyway, since that's probably the better thing to do. Can we merge the build and Linux ones to master, then we can move discussion of the refactor to #46 ?

In that PR, let's discuss the path forward!

**Comment by obelisk** [comment]
Closing this in favour of #44 #45 #46

**Pull request closed without merging**